jdaviz/configs/default/plugins/plot_options/plot_options.py

```diff
@@ -600,7 +594,18 @@ def _update_stretch_histogram(self, msg={}):
                 y_min = max(y_limits.min(), 0)
                 y_max = y_limits.max()
 
-                sub_data = comp.data[y_min:y_max, x_min:x_max].ravel()
+                arr = comp.data[y_min:y_max, x_min:x_max]
+
+                size = arr.shape[0] * arr.shape[1]
+                if size > 400**2:
+                    xstep = max(1, round(arr.shape[1] / 400))
+                    ystep = max(1, round(arr.shape[0] / 400))
+                    arr = arr[::ystep, ::xstep]

```

Comment:
@pllim - this code block essentially exists in three places in jdaviz now (compass and twice here).  Do you think it's worth moving somewhere reusable?  Maybe in utils?

Reply:
Not sure. If you want this to turn into utility function, then you have to make sure that function is not making accidental copies (everything must pass by reference where it matters) and maybe a way to adjust sampling factor and maybe even indices (for generic ndim support). Something to think about.

Reply:
Do you want to refactor in this PR? If so, I will wait before reviewing.

Reply:
I had some of those same concerns, so let's leave this as-is for now and consider moving the logic up higher if we need it anywhere else in the future.

Reply:
I have a bad feeling that we might have to revisit this as more and more new analysis stuff get requested. Should we open a ticket pre-emptively for the future?